File: hosted/site/src/lib/manifest.ts
// The normalized manifest hosted/discovery stores, as written by
// product/template/normalize.go. Only types: the site validates nothing,
// it shows what the validator already accepted.

export type NormalizedLimits = { cpu: number | null; memory_bytes: number | null };

export type NormalizedSource =
  | { type: "image"; image: string; tag: string | null }
  | {
      type: "dockerfile" | "railpack";
      repo: string;
      ref: string | null;
      dockerfile: string | null;
    };

export type NormalizedInput = {
  key: string;
  type: "domain" | "text" | "number" | "choice" | "secret" | "store";
  label: string;
  help?: string;
  required: boolean;
  default?: string | number;
  pattern?: string;
  min?: number;
  max?: number;
  options?: string[];
  generate?: number;
};

export type NormalizedApp = {
  key: string;
  name: string;
  source: NormalizedSource;
  port: number | null;
  health: string | null;
  domains: { host: string; port: number }[];
  attach: { kind: "database" | "store"; key: string; bucket: string | null; prefix: string }[];
  env: Record<string, string>;
  limits: NormalizedLimits | null;
  scale: number | null;
  spread: boolean;
  autoscale: { min: number; max: number; cpu: number } | null;
  internal_host: string;
};

export type NormalizedManifest = {
  schema_version: 1;
  min_cubeship: string | null;
  project: string;
  environment: string;
  inputs: NormalizedInput[];
  databases: {
    key: string;
    name: string;
    engine: string;
    version: string | null;
    username: string | null;
    database: string | null;
    expose: number | null;
    limits: NormalizedLimits | null;
    port: number | null;
  }[];
  stores: {
    key: string;
    name: string;
    version: string | null;
    buckets: string[];
    limits: NormalizedLimits | null;
  }[];
  apps: NormalizedApp[];
};

File: product/template/manifest.go
package template

// Manifest is a file as decoded, before normalizing. An empty string is
// a field that was not given; decode refuses an empty one that was.
type Manifest struct {
	Version     int
	MinCubeship string
	Project     string
	Environment string
	Inputs      []Input
	Databases   []Database
	Stores      []Store
	Apps        []App
}

type Input struct {
	Key, Type, Label, Help string
	Required               bool
	Default                any // a string, or a float64 for a number input
	Pattern                string
	Min, Max               *float64
	Options                []string
	Generate               *int
}

type Limits struct {
	CPU    *float64
	Memory *string
}

type Database struct {
	Key, Name, Engine, Version, Username, Database string
	// Expose is nil for internal only, 0 for "pick a port", N for that port.
	Expose *int
	Limits *Limits
}

type Store struct {
	Key, Name, Version string
	Buckets            []string
	Limits             *Limits
}

type Domain struct {
	Host string
	Port *int
}

type Attach struct {
	Database, Store, Bucket, Prefix string
}

type Autoscale struct {
	Min *int
	Max int
	CPU float64
}

type App struct {
	Key, Name, Image, Tag, Repo, Ref, Build, Dockerfile string
	Port                                                *int
	Health                                              string
	Domains                                             []Domain
	Attach                                              []Attach
	Env                                                 map[string]string
	Limits                                              *Limits
	Scale                                               *int
	Spread                                              *bool
	Autoscale                                           *Autoscale
	Volumes                                             []Volume
}

// Volume is a directory inside an app's container that outlives it.
type Volume struct {
	Path string
}
